File: Longest_Substring_Without_Repeating_Characters_/cpp.d/longest_substring.cpp
#include <vector>
#include <stdio.h>
#include <iostream>
#include <algorithm>
#include <unordered_map>


typedef std::unordered_map<std::string, int> expected_result;

class LongestSubstring {
public:
    int lengthOfLongestSubstring(std::string s){
        std::array <int, 256> last_index;
        last_index.fill(-1);
        int n = s.length();
        int max_substr_len = 0;
        int i = 0;

        for (int j=0; j<n; j++){
            i = std::max(i, last_index[(int)(s[j])] + 1);
            max_substr_len = std::max(max_substr_len, j - i + 1);
            last_index[(int)(s[j])] = j;
        }
        return max_substr_len;
    }

    bool function_validate(std::string s, int res){
      printf("Checking string -> %s\n", s.c_str());
      bool bool_result = this->lengthOfLongestSubstring(s) == res;
      if (bool_result){
        std::cout << "string " <<
        s << " yields correct result -> " <<
        res << "\n\n" ;
      } else {
        std::cout << "bool result no likey\n\n";
      }
    return bool_result;
    }
};

int main(){
  LongestSubstring ls;
  expected_result results({
    {"abcabcbb", 3},
    {"bbbbb", 1},
    {"pwwkew", 3},
    {"abcdefg", 7},
    {"aab", 2},
    {"dvdf", 3}
  });
  for (auto x: results){
    ls.function_validate(x.first, x.second);
  }
  return 0;
}

File: Longest_Substring_Without_Repeating_Characters/js.d/longest_substring.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function(s){
    var lastIndex = Array(256).fill(-1);
    const n = s.length;
    var maxSubstrLen = 0;
    var i = 0;

    for (let j=0; j<n; j++){
        i = Math.max(i, lastIndex[s[j].charCodeAt()] + 1);
        maxSubstrLen = Math.max(maxSubstrLen, j - i + 1);
        lastIndex[s[j].charCodeAt()] = j;
    }
    return maxSubstrLen;
};

/***********Test-Driver************
***********************************/
/**
 * @param {string} s
 * @param {int} res
 * @return {bool}
 */
var functionValidate = function(s, res){
    var resValid = lengthOfLongestSubstring(s) == res;
    if (resValid){
      console.log(`String "${s}" showing correct result => ${res}`);
    } else {
      console.log(`[Error]: Incorrect result for "${s}" => ${res}`);
    }
    return resValid;
}

const stringResMap = {
  "abcabcbb": 3,
  "bbbbb": 1,
  "pwwkew": 3,
  "abcdefg": 7,
  "aab": 2,
  "dvdf": 3
}

for (const [key, value] of Object.entries(stringResMap)) {
  functionValidate(key, value);
}
/***********End-Test************
********************************/
